[IMP] web: property modal design

chgo-odoo opened the pull request:
The property modal was recently introduced but the design was implemented in a short time and could be improved.

Since the form view will have a fresh new design, the purpose of this task is to make the modal fits with that new vibe.

Note that the overall modal design is not being modified since it's used in other modules and fits already with the UI.

task-3084756

Requires : https://github.com/odoo/enterprise/pull/40706

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/119353).

chgo-odoo commented:
@stefanorigano Here is the PR related to the modal property design if you wanna have a look at it.

stefanorigano commented:
Good, thanks @chgo-odoo .
@Brieuc-brd , could you have a first look? thanks

Brieuc-brd commented on addons/web/static/src/views/fields/properties/property_definition.xml:
This bug seems to be already in master, but it would be nice to fix it in your redesign :
Since the inputs have a white background in Community, the Field Type icon is hidden. Playing with `z-index` could fix the issue.
![Capture d’écran 2023-04-24 à 10 05 48](https://user-images.githubusercontent.com/80679690/233936836-3903a59d-c826-4f45-969c-9f849be2258c.png)
<img width="546" alt="Capture d’écran 2023-04-24 à 10 06 22" src="https://user-images.githubusercontent.com/80679690/233936840-6ab485fe-92e0-4d1e-9f18-bb6846f25963.png">

Brieuc-brd reviewed (changes requested):
Hello @chgo-odoo , thanks for your work 👍 
Some comments below 👇

stefanorigano commented on addons/web/static/src/scss/bootstrap_overridden.scss:
```diff
@@ -212,6 +212,7 @@ $nav-pills-link-active-bg: $o-brand-primary !default;
 // Popovers
 
 $popover-border-radius: $border-radius !default;
+$popover-border-color: $border-color;
```
Yes, I agree

chgo-odoo commented:
@brd

> Hello @chgo-odoo , thanks for your work 👍 Some comments below 👇

Hi @Brieuc-brd , thanks for the feedback !

Requested changes have been done + I fixed some class order issues and removed some useless classes.

Brieuc-brd reviewed (changes requested):
Thanks @chgo-odoo 👍 
Some remarks below.

chgo-odoo commented:
@Brieuc-brd changes done

Brieuc-brd reviewed (approved):
@chgo-odoo LGTM 👍 
@stefanorigano Can you check ?

stefanorigano reviewed (changes requested):
Very good @chgo-odoo , thanks 👍 

I suggest one little improvement:
Replace the `<table>`'s  `m-3` class with `mx-3 mt-4`... 
![image](https://user-images.githubusercontent.com/8558351/234294009-0757a6b9-fe24-4e72-9222-cfd4ead45f94.png)

What do you think ?

chgo-odoo commented:
> Very good @chgo-odoo , thanks 👍
> 
> I suggest one little improvement: Replace the `<table>`'s `m-3` class with `mx-3 mt-4`... ![image](https://user-images.githubusercontent.com/8558351/234294009-0757a6b9-fe24-4e72-9222-cfd4ead45f94.png)
> 
> What do you think ?

This brings definitely more whitespace inside the modal, thanks for the suggestion 👍

@stefanorigano changes done and ready for the next stage, to confirm, I didn't add `border-radius: $border-radius` in the CSS since milk will set a border-radius for the element.

chgo-odoo commented:
@stefanorigano Could you have a look at this PR again? Thank you 👍

stefanorigano commented on addons/web/static/src/views/fields/properties/property_definition.scss:
```diff
@@ -6,21 +6,9 @@
-        }
+    .o_modal_container {
+        min-width: 425px;
+        max-width: 550px;
```
Arbitrary and veeeeery specific 😅
Isn't there something we can refer too rather than force a specific value? Normally you should ask yourself: "would this value still be correct if we change the base font-size?" or grid-gutter? or the base spacing?"

stefanorigano commented on addons/web/static/src/views/fields/properties/property_definition.scss:
```diff
@@ -45,7 +33,7 @@
 
-    td {
+    td:not(:last-of-type) {
         padding-bottom: 15px;
```
since we're here, better to replace `15px` with something like `map-get($spacers, 3)` or smthg...

stefanorigano commented on addons/web/static/src/views/fields/properties/property_definition_selection.scss:
```diff
@@ -1,10 +1,5 @@
-    .fa-star,
-    .fa-star-o {
-        color: gold;
+    .fa-star{
```
```suggestion
    .fa-star {
```
😉

stefanorigano commented on addons/web/static/src/views/fields/properties/property_definition.xml:
```diff
@@ -2,49 +2,54 @@
+                    <i class="fa fa-trash"/> Delete
+                </button>
             </div>
-            <table class="m-3">
```
Replacing  `table` with `div` is making fail several JS tests.. are you sure that this change is necessary?

stefanorigano commented on addons/web/static/src/views/fields/properties/property_definition.xml:
```diff
@@ -71,14 +76,14 @@
-                    <td class="o_td_label align-top pe-2">
+                <div t-if="state.propertyDefinition.type === 'selection'" class="d-flex align-items-center">
+                    <div class="w-25 align-self-start me-2">
                         <b>Values</b>
```
Ideally I would like this card to inherit the formView design. Could you use the same structure and classes?

stefanorigano reviewed (changes requested):
Thanks @chgo-odoo , good 👍 
Some remarks for you.

stefanorigano commented on addons/web/static/src/views/fields/properties/property_definition.xml:
```diff
@@ -71,14 +75,14 @@
-                    </td>
-                    <td>
+                <div t-if="state.propertyDefinition.type === 'selection'" class="d-contents mb-3 mb-sm-0">
+                    <div class="text-900">
```
Shouldn't the text-color be controlled by the `o_form_label` class?
If I'm wrong and you still need the `text-900` class I propose to apply it to the label directly

stefanorigano commented on addons/web/static/src/views/fields/properties/property_definition.xml:
```diff
@@ -71,14 +75,14 @@
+                    <div class="text-900">
+                        <label class="o_form_label">Values</label>
+                    </div>
+                    <div class="text-900">
```
I think we can remove this `<div>` and directly apply the `text-900` class to the `web.PropertyDefinitionSelection` component.  Could you check?

stefanorigano commented on addons/web/static/src/views/fields/properties/property_definition.xml:
```diff
@@ -87,15 +91,13 @@
-                    <td>
+                    </div>
+                </div>
+                <div t-if="state.propertyDefinition.type === 'tags'" class="d-contents mb-3 mb-sm-0">
```
Same as before: `text-900`applied on labels and components directly + remove the`<div>`'s (if possible).
Could you check for other occurrences using the same structure too? Like `DomainSelector` etc..

stefanorigano reviewed (commented):
Very good @chgo-odoo , thanks.
Some remarks... almost there, courage 💪

chgo-odoo commented on addons/web/static/src/views/fields/properties/property_definition.scss:
```diff
@@ -1,26 +1,10 @@
-        }
+    .o_modal_container {
+        min-width: calc(#{$o-form-view-sheet-max-width} / 3);
+        max-width: calc(#{$o-form-view-sheet-max-width} / 2);
```
Just checked locally, it seems like it's working well !

stefanorigano commented on addons/web/static/src/views/fields/properties/property_definition.xml:
```diff
@@ -71,128 +69,100 @@
-            </table>
+                    class="o_field_property_definition_value d-contents mb-3 mb-sm-0">
+                    <label t-att-for="getUniqueDomID('default')" class="o_form_label align-self-center text-900">
+                        <b t-if="state.propertyDefinition.type === 'boolean'">
```
Since the `<b>` tag remain untouched, it's better to keep it outside the condition.
Scenario: let's assume that in the future we want to replace the label using xpath, using your implementation we'll have to replace both `<b>`... this way just one ;) 

```suggestion
                        <b>
                            <t t-if="state.propertyDefinition.type === 'boolean'">Default State</t>
                            <t t-else="">Default Value</t>
                         </b>
                                
```

stefanorigano reviewed (changes requested):
Just a minor remark @chgo-odoo 👍 
For the rest LGTM. You can remove the draft state and proceed with the tech. review

chgo-odoo commented:
@stefanorigano Forgot to ping you on the [Enterprise PR](https://github.com/odoo/enterprise/pull/40706), could you have a look at it ? Should be fast since we don't modify much things in there, thanks ! 👍

chgo-odoo commented:
Hi @std-odoo,

Here are two PR's related to the Property Modal Design. Could you please have a look at these when you have an availability?

Appreciate it 👍

std-odoo commented on addons/web/static/src/views/fields/properties/property_definition.xml:
```diff
@@ -1,50 +1,48 @@
+            <div class="o_modal_container p-3 o_inner_group grid">
+                <div class="d-contents mb-3 mb-sm-0">
+                    <label class="o_form_label align-self-center text-900">Label</label>
+                    <h6 t-if="props.readonly" t-out="state.propertyDefinition.string" class="text-truncate"/>
```
can you split that line ?

std-odoo commented on addons/web/static/src/views/fields/properties/property_definition.xml:
```diff
@@ -1,50 +1,48 @@
+                    />
+                </div>
+                <div class="d-contents mb-3 mb-sm-0">
+                    <label t-att-for="getUniqueDomID('type')" class="o_form_label align-self-center text-900">Field Type</label>
```
```suggestion
                    <label t-att-for="getUniqueDomID('type')"
                        class="o_form_label align-self-center text-900">
                        Field Type
                    </label>
```

std-odoo commented on addons/web/static/src/views/fields/properties/property_definition.xml:
```diff
@@ -54,7 +52,7 @@
                             <t t-set-slot="toggler">
                                 <div class="o_input_dropdown w-100 o_field_property_dropdown">
-                                    <img t-attf-src="/web/static/src/views/fields/properties/icons/{{state.propertyDefinition.type}}.png"/>
+                                    <img t-attf-src="/web/static/src/views/fields/properties/icons/{{state.propertyDefinition.type}}.png" class="position-relative z-index-1 me-n4"/>
```
same here

std-odoo commented on addons/web/static/src/views/fields/properties/property_definition.xml:
```diff
@@ -71,128 +69,98 @@
+                    />
+                </div>
+                <div t-if="state.propertyDefinition.type === 'tags'" class="d-contents mb-3 mb-sm-0">
+                    <label t-att-for="getUniqueDomID('tags')" class="o_form_label text-900">Tags</label>
```
and here

std-odoo reviewed (commented):
@chgo-odoo Hi :) 

Look almost good to me, can you just avoid long line ? (80-90 chars max)

Thanks for your work :)

chgo-odoo commented:
Hi @std-odoo 👋

Thanks for the feedback ! Requested changes have been done, how does it look now?

chgo-odoo commented:
Good afternoon @tde-banana-odoo 👋

I've made some changes to the modal (rebased the branch to get a commit related to the modal design) and fixed some alignment issues.

Could you please check this PR and [the Enterprise one](https://github.com/odoo/enterprise/pull/40706) ?

Many thanks already !

Cheers 🙌

chgo-odoo commented:
Hi @std-odoo, hope you had a nice weekend ! 👋

LNA told me that you might be able to help making this PR move forward 🚀 

As I said to TDE, I've made some changes to the modal (rebased the branch to get a commit related to the modal design) and fixed some alignment issues.

If you could review this PR and [the Enterprise one](https://github.com/odoo/enterprise/pull/40706) once again, it would be much appreciated !

Thanks !

awa-odoo commented on addons/web/static/src/views/fields/properties/property_definition.xml:
```diff
@@ -71,66 +74,48 @@
+                        onTagsChange.bind="onTagsChange"/>
+                </div>
+                <div t-if="state.propertyDefinition.type === 'many2one' || state.propertyDefinition.type === 'many2many'" class="o_field_property_definition_model d-contents mb-3 mb-sm-0">
+                <label t-att-for="getUniqueDomID('model')" class="o_form_label align-self-center text-900">Model</label>
+                <span t-if="props.readonly">
+                    <t t-if="state.resModelDescription" t-out="state.resModelDescription"/>
+                </span>
+                <ModelSelector
+                    t-else=""
+                    id="getUniqueDomID('model')"
+                    value="state.resModelDescription"
+                    onModelSelected.bind="onModelChange"
+                />
```
Aren't we missing an indentation level here?

awa-odoo commented on addons/web/static/src/views/fields/properties/property_definition.xml:
```diff
@@ -1,50 +1,52 @@
-                        t-on-click="() => this.onPropertyMove('down')">
-                        <i class="oi oi-chevron-down"/>
+                        t-attf-class="btn btn-link ps-0 {{this.isFirst ? 'disabled' : ''}}"
+                        t-on-click="() => this.onPropertyMove('up')" title="Move this Property up">
```
As you have new titles, I think this requires an update of the .pot file.

awa-odoo reviewed (commented):
Hello @chgo-odoo 

One minor comment and one more important about translations.

Can you take a look?

Thanks!

chgo-odoo commented:
Hi @awa-odoo, thanks for reviewing ! 

Requested changes have been done, if you could have a look at it? Thanks !

awa-odoo commented:
@robodoo r+

robodoo commented:
@chgo-odoo @awa-odoo linked pull request(s) odoo/enterprise#40706 not ready. Linked PRs are not staged until all of them are ready.
